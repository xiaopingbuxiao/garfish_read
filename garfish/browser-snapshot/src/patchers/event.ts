let rawAddEventListener: any;
let rawRemoveEventListener: any;

export class PatchEvent {
  private listenerMap = new Map<string, EventListenerOrEventListenerObject[]>();
  constructor() {}

  public activate() {
    // restore pre event
    this.listenerMap.forEach((listeners, type) =>
      [...listeners].forEach((listener) =>
        window.addEventListener(type, listener),
      ),
    );

    if (!rawAddEventListener || !rawRemoveEventListener) {
      rawAddEventListener = window.addEventListener;
      rawRemoveEventListener = window.removeEventListener;
    }

    window.addEventListener = (
      type: string,
      listener: EventListenerOrEventListenerObject,
      options?: boolean | AddEventListenerOptions,
    ) => {
      const listeners = this.listenerMap.get(type) || [];
      this.listenerMap.set(type, [...listeners, listener]);
      return rawAddEventListener.call(window, type, listener, options);
    };
    window.removeEventListener = (
      type: string,
      listener: EventListenerOrEventListenerObject,
      options?: boolean | AddEventListenerOptions,
    ) => {
      const storedTypeListeners = this.listenerMap.get(type);
      if (
        storedTypeListeners &&
        storedTypeListeners.length &&
        storedTypeListeners.indexOf(listener) !== -1
      ) {
        storedTypeListeners.splice(storedTypeListeners.indexOf(listener), 1);
      }
      return rawRemoveEventListener.call(window, type, listener, options);
    };
  }

  public deactivate() {
    this.listenerMap.forEach((listeners, type) =>
      [...listeners].forEach((listener) =>
        window.removeEventListener(type, listener),
      ),
    );

    // event，在window原型链上，将window上覆盖的代理事件删除即可
    // delete window.removeEventListener;
    // delete window.addEventListener;
    window.removeEventListener = rawRemoveEventListener;
    window.addEventListener = rawAddEventListener;
  }
}
