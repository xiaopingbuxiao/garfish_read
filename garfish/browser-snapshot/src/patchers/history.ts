let rawPushState: any;
let rawReplaceState: any;

export class PatchHistory {
  public activate() {
    if (!rawPushState || !rawReplaceState) {
      rawPushState = window.history.pushState;
      rawReplaceState = window.history.replaceState;
    }
  }

  public deactivate() {
    window.history.pushState = rawPushState;
    window.history.replaceState = rawReplaceState;
  }
}
