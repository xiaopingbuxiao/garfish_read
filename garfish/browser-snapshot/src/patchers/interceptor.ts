export type SnapshotDiff = {
  created: Snapshot;
  removed: Snapshot;
};

export function isStyledComponentsLike(element: HTMLStyleElement) {
  // A styled-components liked element has no textContent but keep the rules in its sheet.cssRules.
  return (
    element instanceof HTMLStyleElement &&
    !element.textContent &&
    element.sheet?.cssRules.length
  );
}

export class Snapshot {
  constructor(public arrDoms: Array<HTMLElement>) {
    this.arrDoms = arrDoms;
  }

  static take(target: HTMLElement): Snapshot;
  static take<T extends Element = Element>(
    nodeList: HTMLCollectionOf<T>,
  ): Snapshot;
  static take(nodeList?: NodeList): Snapshot;
  static take(target: HTMLElement | HTMLCollection | NodeList = document.head) {
    let list: Array<HTMLElement>;
    if ((target as HTMLElement).childNodes) {
      list = Array.prototype.slice.call((target as HTMLElement).childNodes);
    } else {
      list = Array.prototype.slice.call(target);
    }
    return new Snapshot(list);
  }

  diff(s: Snapshot): SnapshotDiff {
    if (!s) {
      return {
        created: new Snapshot([]),
        removed: new Snapshot([]),
      };
    }

    return {
      created: new Snapshot(
        this.arrDoms.filter((d) => s.arrDoms.indexOf(d) === -1),
      ),
      removed: new Snapshot(
        s.arrDoms.filter((d) => this.arrDoms.indexOf(d) === -1),
      ),
    };
  }
}

export class Interceptor {
  dynamicStyleSheetElementSet: Set<HTMLStyleElement>;
  styledComponentCSSRulesMap: WeakMap<HTMLStyleElement, CSSRuleList>;
  constructor(public dom: HTMLElement = document.head) {
    this.dom = dom;
    this.dynamicStyleSheetElementSet = new Set<HTMLStyleElement>();
    this.styledComponentCSSRulesMap = new WeakMap<
      HTMLStyleElement,
      CSSRuleList
    >();
  }

  add(from: Snapshot): void;
  add(created: Snapshot, removed: Snapshot): void;
  add(createdOrSnapshot: Snapshot, removed?: Snapshot) {
    let created: Snapshot;
    if (!removed) {
      const diff = Snapshot.take(this.dom).diff(createdOrSnapshot);
      created = diff.created;
      removed = diff.removed;
    } else {
      created = createdOrSnapshot;
    }
    created.arrDoms.reduce((prev, val) => {
      prev.appendChild(val);
      if (val instanceof HTMLStyleElement) {
        const cssRules = this.styledComponentCSSRulesMap.get(val);
        if (cssRules && cssRules.length) {
          for (let i = 0; i < cssRules.length; i++) {
            const cssRule = cssRules[i];
            // re-insert rules for styled-components element
            val.sheet?.insertRule(cssRule.cssText, val.sheet?.cssRules.length);
          }
        }
      }
      return prev;
    }, this.dom);
    removed.arrDoms.reduce((prev, val) => {
      prev.removeChild(val);
      return prev;
    }, this.dom);
  }

  remove(to: Snapshot): void;
  remove(created: Snapshot, removed: Snapshot): void;
  remove(createdOrSnapshot: Snapshot, removed?: Snapshot) {
    let created: Snapshot;
    if (!removed) {
      const diff = Snapshot.take(this.dom).diff(createdOrSnapshot);
      created = diff.created;
      removed = diff.removed;
    } else {
      created = createdOrSnapshot;
    }
    created.arrDoms.reduce((prev, val) => {
      if (
        val instanceof HTMLStyleElement &&
        isStyledComponentsLike(val) &&
        val?.sheet?.cssRules
      ) {
        this.styledComponentCSSRulesMap.set(val, val.sheet.cssRules);
      }
      prev.removeChild(val);
      return prev;
    }, this.dom);
    removed.arrDoms.reduce((prev, val) => {
      prev.appendChild(val);
      return prev;
    }, this.dom);
  }
}
