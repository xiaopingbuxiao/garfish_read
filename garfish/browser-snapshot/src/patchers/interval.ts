const rawInterval = window.setInterval;
const rawClearInterval = window.clearInterval;

export class PatchInterval {
  private intervals: Array<number> = [];
  constructor() {}

  public activate() {
    // @ts-ignore
    window.setInterval = (
      handler: Function,
      timeout?: number,
      ...args: any[]
    ) => {
      const intervalId = rawInterval(handler, timeout, ...args);
      this.intervals = [...this.intervals, intervalId];
      return intervalId;
    };

    // @ts-ignore
    window.clearInterval = (intervalId: number) => {
      this.intervals = this.intervals.filter((id) => id !== intervalId);
      return rawClearInterval(intervalId);
    };
  }

  public deactivate(_clearEffects?: boolean) {
    if (_clearEffects) {
      this.intervals.forEach((id) => window.clearInterval(id));
    }
    window.setInterval = rawInterval;
    window.clearInterval = rawClearInterval;
  }
}
