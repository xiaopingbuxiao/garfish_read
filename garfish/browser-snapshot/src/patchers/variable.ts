const hasOwnProperty = Object.prototype.hasOwnProperty;
function hasOwn(obj: any, key: PropertyKey): boolean {
  return hasOwnProperty.call(obj, key);
}

export class PatchGlobalVal {
  public snapshotOriginal = new Map();
  private snapshotMutated: any = new Map();
  private whiteList: Array<PropertyKey> = [
    'location',
    'addEventListener',
    'removeEventListener',
    'webpackJsonp',
  ];
  // ,'addEventListener','removeEventListener','clearInterval','setInterval','webkitStorageInfo'
  constructor(
    public targetToProtect: any = typeof window !== 'undefined'
      ? window
      : globalThis,
    public protectVariable: Array<PropertyKey> = [],
  ) {
    this.targetToProtect = targetToProtect;
    this.protectVariable = protectVariable;
    // this.whiteList = [...this.whiteList, ...GarConfig.protectVariable!];
    this.whiteList = [...this.whiteList, ...protectVariable];
  }

  protected safeIterator(fn: Function) {
    // Skip the variables not traverse
    // Do not include a symbol
    for (const i in this.targetToProtect) {
      if (this.whiteList.indexOf(i) !== -1) {
        continue;
      }
      const prop = Object.getOwnPropertyDescriptor(this.targetToProtect, i);
      if (!prop || !prop.writable) {
        continue;
      }
      if (hasOwn(this.targetToProtect, i)) {
        fn(i);
      }
    }
  }

  // 1.Trigger hooks, life cycle willActivate enabled (going to)
  // 2.Will disable the current group of other box, and triggers the switch life cycle
  // 3.The current window object properties for caching
  // 4.Restore the sandbox side effects during operation
  public activate() {
    // Recorded before the global environment, restore side effects of a variable
    this.safeIterator((i: string) => {
      this.snapshotOriginal.set(i, this.targetToProtect[i]);
    });

    this.snapshotMutated.forEach((val, mutateKey) => {
      this.targetToProtect[mutateKey] = this.snapshotMutated.get(mutateKey);
    });
  }

  // 1.Restore the sandbox during startup variables change, record the change record
  // 2.Restore the sandbox during startup to delete variables, record the change record
  public deactivate() {
    const deleteMap: any = {};
    const updateMap: any = {};
    const addMap: any = {};

    // Restore the sandbox before running Windows properties of environment, and difference value for caching
    this.safeIterator((normalKey: string) => {
      if (
        this.snapshotOriginal.get(normalKey) !==
        (this.targetToProtect[normalKey] as any)
      ) {
        this.snapshotMutated.set(normalKey, this.targetToProtect[normalKey]); // deleted key will be defined as undefined on
        this.targetToProtect[normalKey] = this.snapshotOriginal.get(normalKey); // || this.targetToProtect[i]

        // Collection of delete, modify variables
        if (this.targetToProtect[normalKey] === undefined) {
          addMap[normalKey] = this.snapshotMutated.get(normalKey);
        } else {
          updateMap[normalKey] = this.snapshotMutated.get(normalKey);
        }
      }
      this.snapshotOriginal.delete(normalKey);
    });

    this.snapshotOriginal.forEach((val, deleteKey) => {
      this.snapshotMutated.set(deleteKey, this.targetToProtect[deleteKey]);
      this.targetToProtect[deleteKey] = this.snapshotOriginal.get(deleteKey);
      deleteMap[deleteKey] = this.targetToProtect[deleteKey];
    });

    // For developers, let them know clear what side effects of a variable
    // channel.emit('sandbox-variable', {
    //   update: updateMap,
    //   removed: deleteMap,
    //   add: addMap,
    // });
  }
}
