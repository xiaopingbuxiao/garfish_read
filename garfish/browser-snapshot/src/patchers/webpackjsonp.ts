declare global {
  interface Window {
    webpackJsonp?: any[];
  }
}

export class PatchWebpackJsonp {
  preWebpackJsonp?: any[];

  currentWebpackJsonp?: any[];

  public activate() {
    this.preWebpackJsonp = window.webpackJsonp;
    window.webpackJsonp = this.currentWebpackJsonp;
  }

  public deactivate() {
    this.currentWebpackJsonp = window.webpackJsonp;
    window.webpackJsonp = this.preWebpackJsonp;
  }
}
