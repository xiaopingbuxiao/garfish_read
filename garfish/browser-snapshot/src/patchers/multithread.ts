import { Snapshot } from './interceptor';
export class MultithreadInterceptor {
  snapshotBefore: Snapshot;
  targetToProtect: HTMLElement;
  uuid: string;
  constructor() {
    this.uuid = '' + -new Date();
    this.targetToProtect.setAttribute('gar', this.uuid);
  }
  beforeAdd() {
    this.snapshotBefore = Snapshot.take(document.getElementsByTagName('style'));
  }
  afterAdd() {
    const now = Snapshot.take(document.getElementsByTagName('style'));
    const diff = this.snapshotBefore.diff(now);
    diff.created.arrDoms.forEach((styleElement: HTMLStyleElement) => {
      const styleSheet: CSSStyleSheet = styleElement.sheet as CSSStyleSheet;
      if (!styleSheet.cssRules) {
        return;
      }
      for (
        let i = 0, rule: CSSStyleRule;
        (rule = styleSheet.cssRules[i] as CSSStyleRule);
        i++
      ) {
        rule.selectorText += `[gar=${this.uuid}] `;
      }
    });
  }
}
