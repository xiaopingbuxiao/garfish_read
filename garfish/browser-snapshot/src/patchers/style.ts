import { Interceptor, Snapshot, SnapshotDiff } from './interceptor';

interface EffectCt {
  src: string;
  outerHTML: string;
  content: string;
}

interface EffectMap {
  style: Array<EffectCt>;
  script: Array<EffectCt>;
  other: Array<EffectCt>;
}

export class PatchStyle {
  public headInterceptor: Interceptor;
  private domSnapshotBefore!: Snapshot;
  private domSnapshotMutated!: SnapshotDiff | null;

  constructor() {
    this.headInterceptor = new Interceptor(document.head);
  }

  public activate() {
    // 记录当前dom节点、恢复之前dom节点副作用
    this.domSnapshotBefore = Snapshot.take();
    if (this.domSnapshotMutated)
      this.headInterceptor.add(
        this.domSnapshotMutated.created,
        this.domSnapshotMutated.removed,
      );
  }

  public deactivate() {
    // 恢复沙盒运行前dom节点环境，并将差异值进行缓存
    const domSnapshot = Snapshot.take();
    this.domSnapshotMutated = domSnapshot.diff(this.domSnapshotBefore);
    if (!this.domSnapshotMutated) return;
    this.headInterceptor.remove(
      this.domSnapshotMutated.created,
      this.domSnapshotMutated.removed,
    );
  }

  private formateCtx(arrDoms: Array<HTMLElement>) {
    const effectMap: EffectMap = {
      style: [],
      script: [],
      other: [],
    };

    arrDoms.forEach((dom) => {
      let type: 'other' | 'style' | 'script' = 'other';
      if (/css/.test((dom as any).type)) type = 'style';
      if (/javascript/.test((dom as any).type)) type = 'script';
      effectMap[type].push({
        src: (dom as any).src,
        outerHTML: dom.outerHTML,
        content: dom.innerText,
      });
    });

    return effectMap;
  }
}
