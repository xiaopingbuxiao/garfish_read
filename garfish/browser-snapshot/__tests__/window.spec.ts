import { PatchGlobalVal } from '../src/patchers/variable';

declare global {
  interface Window {
    [index: string]: any;
  }
}

// 1.Can remove the sand box running normal test box during the side effects of the variables on the window global variables
// 2.Test sandbox can protect properly named variables in the closure will not remove the sandbox
describe('snapshot sandbox window', () => {
  it('test normal obj', () => {
    const obj: any = {
      update: 'update value',
      name: 'add delete val',
      nObj: { name: 'hello' },
    };

    const nVal = new PatchGlobalVal(obj);
    nVal.activate();

    obj.a = 'hello';
    obj.update = 'new value';
    delete obj.name;

    nVal.deactivate();

    expect(obj).toMatchSnapshot();

    nVal.activate();
    expect(obj).toMatchSnapshot();
  });

  it('Add a variable test', () => {
    const PatchGlobal = new PatchGlobalVal(window);

    const fn = function a() {};
    const ar = [];
    const ob = { name: 'obj' };

    // snapshot
    PatchGlobal.activate();

    window.fn = fn;
    window.nm = 123;
    window.st = 'string';
    window.ar = ar;
    window.bl = true;
    window.nl = null;
    (window as any)[1] = 1234;
    window.ob = ob;

    // Remove the side effects
    PatchGlobal.deactivate();

    expect(window.fn).toBe(undefined);
    expect(window.nm).toBe(undefined);
    expect(window.st).toBe(undefined);
    expect(window.ar).toBe(undefined);
    expect(window.bl).toBe(undefined);
    expect(window.nl).toBe(undefined);
    expect(window[1]).toBe(undefined);
    expect(window.ob).toBe(undefined);

    // restore side effects
    PatchGlobal.activate();

    expect(window.fn).toBe(fn);
    expect(window.nm).toBe(123);
    expect(window.st).toBe('string');
    expect(window.ar).toBe(ar);
    expect(window.bl).toBe(true);
    expect(window.nl).toBe(null);
    expect(window[1]).toBe(1234);
    expect(window.ob).toBe(ob);
  });

  it('Test to delete variables', () => {
    const PatchGlobal = new PatchGlobalVal(window);
    const fn = function a() {};
    const ar = [];
    const ob = { name: 'obj' };
    window.fn = fn;
    window.nm = 123;
    window.st = 'string';
    window.ar = ar;
    window.bl = true;
    window.nl = null;
    (window as any)[1] = 1234;
    window.ob = ob;

    // snapshot
    PatchGlobal.activate();

    delete window.fn;
    delete window.nm;
    delete window.st;
    delete window.ar;
    delete window.bl;
    delete window.nl;
    delete (window as any)[1];
    delete window.ob;

    // Remove the side effects
    PatchGlobal.deactivate();

    expect(window.fn).toBe(fn);
    expect(window.nm).toBe(123);
    expect(window.st).toBe('string');
    expect(window.ar).toBe(ar);
    expect(window.bl).toBe(true);
    expect(window.nl).toBe(null);
    expect(window[1]).toBe(1234);
    expect(window.ob).toBe(ob);

    // restore side effects
    PatchGlobal.activate();

    expect(window.fn).toBe(undefined);
    expect(window.nm).toBe(undefined);
    expect(window.st).toBe(undefined);
    expect(window.ar).toBe(undefined);
    expect(window.bl).toBe(undefined);
    expect(window.nl).toBe(undefined);
    expect(window[1]).toBe(undefined);
    expect(window.ob).toBe(undefined);
  });
});
