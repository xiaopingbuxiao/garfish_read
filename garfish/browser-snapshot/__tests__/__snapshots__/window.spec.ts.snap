// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot sandbox window test normal obj 1`] = `
Object {
  "a": undefined,
  "nObj": Object {
    "name": "hello",
  },
  "name": "add delete val",
  "update": "update value",
}
`;

exports[`snapshot sandbox window test normal obj 2`] = `
Object {
  "a": "hello",
  "nObj": Object {
    "name": "hello",
  },
  "name": undefined,
  "update": "new value",
}
`;
